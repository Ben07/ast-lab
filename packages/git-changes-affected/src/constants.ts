export enum GIT_OPERATION {
  change = 0,
  new,
  delete,
  rename
};