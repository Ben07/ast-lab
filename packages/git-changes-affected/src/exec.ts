import util from 'util';
import { exec } from 'child_process';

const pExec = util.promisify(exec);

export default pExec;