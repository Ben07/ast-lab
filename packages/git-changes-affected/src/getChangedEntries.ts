import { parse, ParserOptions } from '@babel/parser';
import { File, SourceLocation } from '@babel/types';
import { extractStats, getPrivateMemberNames } from 'es-stats';
import _debug from 'debug';
import getAbsolutePath from './getAbsolutePath';
import { Entry, MemberRef, MemberRelation } from 'ast-lab-types';
import { Change } from './types';


const debug = _debug('git-changes-affected:entries');

type Declare = MemberRef & { loc: SourceLocation };
/** @todo improve precision of private declaration location */
function locatePrivateDeclares(ast: File): Declare[] {
  return ast.program.body.reduce((locs, d) => {
    const refs = getPrivateMemberNames(d);
    const { loc } = d;
    if (!refs || !refs.length || !loc) { return locs; }
    return locs.concat(
      refs.map(ref => ({
        ...ref,
        loc
      }))
    );
  }, [] as Declare[]);
}

export default function getChangedEntries(changes: Change[], parserOptions?: ParserOptions | null) {
  return changes.reduce((res, { file, content, changed }) => {
    if (!content) {
      return res;
    }
    const filePath = getAbsolutePath(file);
    let ast;
    try { 
      ast = (parse(content, { 
        ...(parserOptions || {}),
        sourceType: 'module'
      }) as File);
    } catch(e) {
      console.warn(`@bable/parser parsing ${filePath} failed! (${e.message}) Parser options: ${JSON.stringify(parserOptions)}`);
      console.warn('File content >>> \n', content);
      return res;
    }
    const stats = extractStats(ast);
    debug(`${file} stats: ${JSON.stringify(stats, null, 2)}`);
    const declareLoc = [
      ...locatePrivateDeclares(ast),
      ...(stats.exports.members as Declare[]),
      ...(stats.imports as Declare[])
    ];
    debug(`${file} declareLoc: ${JSON.stringify(declareLoc, null, 2)}`);
    const exported = new Set(stats.exports.members.map(({ alias }) => alias));
    const affectExports = {} as MemberRelation;
    exported.forEach(name => {
      const expanded = new Set(stats.relations[name] || []);
      expanded.forEach(m => {
        if (m === name) { return; }
        affectExports[m] = [
          ...(affectExports[m] || []),
          name
        ];
        if (stats.relations[m]) {
          stats.relations[m].forEach(r => expanded.add(r));
        }
      });
    });

    let iDeclare = 0;
    const changedExports = changed.reduce(
      (ex, { start: startLine, end: endLine }) => {
        while (iDeclare < declareLoc.length) {
          const { loc, alias } = declareLoc[iDeclare];
          iDeclare++;
          if (!loc) { continue; }
          const { start, end } = loc;
          if (endLine < start.line) { 
            return ex;
          }
          if (startLine <= end.line) {
            if (exported.has(alias)) {
              ex.push(alias);
            } else if (affectExports[alias]) {
              ex = ex.concat(affectExports[alias]);
            }
          }
        }
        return ex;
      }, 
      [] as string[]
    );
    return res.concat(
      Array.from(new Set(changedExports)).map(name => ({
        source: filePath,
        name
      }))
    );
  }, [] as Entry[]);
}