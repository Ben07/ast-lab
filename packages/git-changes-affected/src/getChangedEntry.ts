import { parse, ParserOptions } from '@babel/parser';
import { File, SourceLocation } from '@babel/types';
import { extractStats, getPrivateMemberNames } from 'es-stats';
import _debug from 'debug';
import getAbsolutePath from './getAbsolutePath';
import { Entry, MemberRef, MemberRelation } from 'ast-lab-types';
import { Change } from './types';

const debug = _debug('git-changes-affected:entries');

type Declare = MemberRef & { loc: SourceLocation };
/** @todo improve precision of private declaration location */
function locatePrivateDeclares(ast: File): Declare[] {
  return ast.program.body.reduce((locs, d) => {
    const refs = getPrivateMemberNames(d);
    const { loc } = d;
    if (!refs || !refs.length || !loc) {
      return locs;
    }
    return locs.concat(
      refs.map(ref => ({
        ...ref,
        loc,
      }))
    );
  }, [] as Declare[]);
}

/**
 * Find what declarations does the code line changes belong to.
 * @param changes
 * @param parserOptions `@babel/parser` options
 * @return A list of object contains module absolute path and declaration name
 */
export default function getChangedEntry(
  { file, content, changed }: Change,
  parserOptions?: ParserOptions | null
): Entry[] | null {
  if (!content) {
    return null;
  }
  const filePath = getAbsolutePath(file);
  let ast;
  try {
    ast = parse(content, {
      ...(parserOptions || {}),
      sourceType: 'module',
    }) as File;
  } catch (e) {
    console.warn(
      `@bable/parser parsing ${filePath} failed! (${
        e.message
      }) Parser options: ${JSON.stringify(parserOptions)}`
    );
    return null;
  }
  const stats = extractStats(ast);
  debug(`${file} stats: ${JSON.stringify(stats)}`);
  const declareLoc = [
    ...locatePrivateDeclares(ast),
    ...(stats.exports.members as Declare[]),
    ...(stats.imports as Declare[]),
  ].sort((a, b) => a.loc.start.line - b.loc.start.line);
  debug(`${file} declareLoc: ${JSON.stringify(declareLoc)}`);

  const exported = new Set(stats.exports.members.map(({ alias }) => alias));
  const affectExports = {} as MemberRelation;
  exported.forEach(name => {
    const expanded = new Set(stats.relations[name] || []);
    expanded.forEach(m => {
      if (m === name) {
        return;
      }
      affectExports[m] = [...(affectExports[m] || []), name];
      if (stats.relations[m]) {
        stats.relations[m].forEach(r => expanded.add(r));
      }
    });
  });
  debug(
    `${file} change ${JSON.stringify(changed)}  \n>>>  ${JSON.stringify(
      affectExports
    )}`
  );

  let iDeclare = 0;
  const changedExports = changed.reduce(
    (ex, { start: startLine, end: endLine }) => {
      while (iDeclare < declareLoc.length) {
        const { loc, alias } = declareLoc[iDeclare];
        if (!loc) {
          continue;
        }
        const { start, end } = loc;
        if (endLine < start.line) {
          debug(`${file}:${startLine}-${endLine} X ${start.line}-${end.line}`);
          return ex;
        }
        if (startLine <= end.line) {
          debug(
            `${file}:${startLine}-${endLine} âˆš ${start.line}-${end.line} > ${affectExports[alias]} `
          );
          if (exported.has(alias)) {
            ex.push(alias);
          }
          if (affectExports[alias]) {
            ex = ex.concat(affectExports[alias]);
          }
        } else {
          debug(`${file}:${startLine}-${endLine} X ${start.line}-${end.line}`);
        }
        iDeclare++;
      }
      return ex;
    },
    [] as string[]
  );
  return Array.from(new Set(changedExports)).map(name => ({
    source: filePath,
    name,
  }));
}
