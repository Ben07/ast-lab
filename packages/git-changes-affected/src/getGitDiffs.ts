import exec from './exec';
import { GIT_OPERATION } from './constants';
import { Diff } from './types';
import getRevisionFile from './getRevisionFile';

function justifyOperation(operatelog: string) {
  if (/^new file mode/.test(operatelog)) {
    return GIT_OPERATION.new;
  }
  if (/^delete file mode/.test(operatelog)) {
    return GIT_OPERATION.delete;
  }
  /** @todo confirm rename similarity percentage */
  if (/^similarity index/.test(operatelog)) {
    return GIT_OPERATION.rename;
  }
  return GIT_OPERATION.change;
}
function isLineRemoved(content: string) {
  return content.startsWith('-');
}
function isLineAdded(content: string) {
  return content.startsWith('+');
}

/**
 * @todo get real diff content
 * @param {string} commit 
 */
export function getGitDiffs(commit: string): Diff[] {
  const strOut = exec(`git show ${commit} --format="%N" --first-parent`);
  const diffs = [] as Diff[];
  let lineA = 0;
  let lineB = 0;
  let aChangeStart = null as number | null;
  let bChangeStart = null as number | null;
  const diffLines = strOut.split('\n')
  diffLines.forEach((content, index) => {
    const lastDiff = diffs[diffs.length - 1];
    const aChanges = lastDiff &&lastDiff.source.changed;
    const bChanges = lastDiff &&lastDiff.target.changed;
    const fileHeadMatch = content.match(/^diff --git a\/([^\n\s]+) b\/([^\n\s]+)/);
    if (fileHeadMatch) {
      if (lastDiff) {
        if (
          lastDiff.operation !== GIT_OPERATION.rename && 
          aChangeStart !== null
        ) {
          aChanges.push({ 
            start: aChangeStart,
            end: lineA - 1
          });
        }
        if (bChangeStart !== null) {
          bChanges.push({
            start: bChangeStart, 
            end: lineB - 1
          });
        }
      }
      const [sourceFile, targetFile] = fileHeadMatch.slice(1);
      const operation = justifyOperation(diffLines[index + 1]);
      diffs.push({
        source: {
          file: sourceFile,
          content: operation === GIT_OPERATION.new ? null : getRevisionFile(`${commit}~1`, sourceFile),
          changed: operation === GIT_OPERATION.rename ? [{
            start: 0, 
            end: Infinity
          }]: []
        },
        target: {
          file: targetFile,
          content:  operation === GIT_OPERATION.delete ? null : getRevisionFile(commit, targetFile),
          changed: []
        },
        operation,
      });
      return;
    }
    const chunkHeadMatch = content.match(/@@ -(\d+),\d+ \+(\d+),\d+ @@/);
    if (chunkHeadMatch) {
      lineA = parseInt(chunkHeadMatch[1], 10);
      lineB = parseInt(chunkHeadMatch[2], 10);
      return;
    }
    const isLastLine = index === diffLines.length - 1;
    const isAdded = isLineAdded(content);
    const isRemoved = isLineRemoved(content);
    if (
      lastDiff.operation !== GIT_OPERATION.rename
    ) {
      if (isRemoved && aChangeStart === null) {
        aChangeStart = lineA;
      }
      if (
        (!isRemoved || isLastLine) && aChangeStart !== null
      ) {
        aChanges.push({ 
          start: aChangeStart,
          end: lineA - 1
        });
        aChangeStart = null;
      }
      if (!isAdded) {
        lineA++;
      }
    }
    if (isAdded && bChangeStart === null) {
      bChangeStart = lineB;
    }
    if (
      (!isAdded || isLastLine) && 
      bChangeStart !== null
    ) {
      bChanges.push({
        start: bChangeStart, 
        end: lineB - 1
      });
      bChangeStart = null;
    }
    if (!isRemoved) {
      lineB++;
    }
  });
  return diffs;
}

export default getGitDiffs;