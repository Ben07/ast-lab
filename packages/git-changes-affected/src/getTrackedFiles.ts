import exec from './exec';
import gitRoot from './gitRoot';

const root = gitRoot();

/** @internal */
export default function getTrackedFiles(revision: string = 'HEAD', paths?: string[]) {
  const pathsInCmd = paths && paths.length ? paths.join(' ') : root;
  const raw = exec(`git ls-tree -r ${revision} --name-only --full-name ${pathsInCmd}`);
  return raw.split('\n');
}