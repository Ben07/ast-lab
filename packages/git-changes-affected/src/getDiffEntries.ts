import { getGitDiffs } from './getGitDiffs';
import { GIT_OPERATION } from './constants';
import { Change } from './types';

export default function getDiffEntries(opts: {
  /** The end result revision. Default to `HEAD`. */
  to?: string;
  /** The revision to compare from. Default to one commit before `to` revision. */
  from?: string;
}): {
  befores: Change[];
  afters: Change[];
} {
  const to = opts.to || `HEAD`;
  const from = opts.from || `${to}~1`;
  const diffs = getGitDiffs(to, from);
  const befores = [] as Change[];
  const afters = [] as Change[];
  diffs.forEach(({ source, target, operation }) => {
    if (operation !== GIT_OPERATION.new) {
      befores.push(source);
    }
    if (operation !== GIT_OPERATION.delete) {
      afters.push(target);
    }
  });
  return {
    befores,
    afters,
  };
}
