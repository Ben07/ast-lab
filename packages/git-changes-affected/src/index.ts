// import fs from 'fs';
import path from 'path';
import { getExports, filterDependents, hasExt, resolve, Alias, ModuleDirectory, Exports } from 'get-dependencies';
import { getGitDiffs, GIT_OPERATION } from './getGitDiffs';
import exec from './exec';

type Transform = (raw: string, path: string) => string;

const gitRoot = exec('git rev-parse --show-toplevel').trim();
const DEFAULT_EXTENSIONS = ['.js', '.jsx', '.ts', '.tsx'];

function getAbsolutePath(relativePath: string) {
  return path.resolve(gitRoot, relativePath);
}
function getRelativePath(file: string) {
  return path.isAbsolute(file) ? path.relative(gitRoot, file) : file;
}

function getRevisionFile(revision: string, file: string) {
  // try {
    return exec(`git show ${revision}:${getRelativePath(file)}`);
  // } catch (err) {
  //   console.warn('Can not find ', file, revision);
  //   return '';
  // }
}

function createLoader(revision: string, transform?: Transform) {
  return (file: string) => {
    const raw = getRevisionFile(revision, file);
    return Promise.resolve(transform ? transform(raw, file) : raw);
  };
}

 export function getTrackedFiles(revision: string = 'HEAD', paths?: string[]) {
  const raw = exec(`git ls-tree -r ${revision} --name-only --full-name ${paths &&paths.length ? paths.join(' ') : gitRoot}`);
  return raw.split('\n').slice(0, -1);
}

type GetDiffExportMapOptions = {
  extensions?: string[],
  transform?: Transform
};
export async function getDiffExportMap(commit: string, { extensions, transform }: GetDiffExportMapOptions) {
  const diffs = getGitDiffs(commit);
  const exportMap = new Map() as Exports;
  const staleExportMap = new Map() as Exports;
  let queue = [] as Promise<void>[];
  diffs.forEach(({ target, source, operation }) => {
    if (!hasExt(target, extensions || DEFAULT_EXTENSIONS)) {
      return;
    }
    queue.push((async () => {
      const targetExports = operation === GIT_OPERATION.delete ?
        [] : 
        await getExports(target, {
          loader: createLoader(commit, transform)
        });
      const sourceExports = operation === GIT_OPERATION.new ?
        [] :
        await getExports(source, {
          loader: createLoader(`${commit}~1`, transform)
        });
      const absoluteTargetPath = getAbsolutePath(target);
      if (source === target) {
        exportMap.set(absoluteTargetPath, new Set([
          ...targetExports,
          ...sourceExports
        ]));
      } else {
        exportMap.set(absoluteTargetPath, new Set(targetExports));
        staleExportMap.set(getAbsolutePath(source), new Set(sourceExports));
      }
    })());
  });
  await Promise.all(queue);
  return [exportMap, staleExportMap];
}

class FileListIncludesPlugin {
  files: Set<string>;
  extensions: string[];

	constructor(files: string[], extensions?: string[]) {
    this.files = new Set(files);
    this.extensions = extensions || ['.jsx', '.js', '.ts', '.tsx']
	}

	apply(resolver: any) {
		const target = resolver.ensureHook('resolved');
		resolver
			.getHook('directory')
			.tapAsync("FileListIncludesPlugin", (request: any, resolveContext: any, callback: any) => {
        const filename = request.path;
        const fileExt = this.extensions.find(ext => {
          const file = `${filename}${ext}`;
          const fileRelativePath =  path.relative(gitRoot, file);
          if (this.files.has(fileRelativePath)) {
            if (resolveContext.fileDependencies)
              resolveContext.fileDependencies.add(file);
            resolver.doResolve(
              target,
              {
                ...request,
                path: file
              },
              "existing file: " + file,
              resolveContext,
              callback
            );
            return true;
          }
          return false;
        });
        if (!fileExt) {
          if (resolveContext.missingDependencies)
          resolveContext.missingDependencies.add(filename);
          if (resolveContext.log) resolveContext.log(filename + " doesn't exist");
          return callback();
        }
			});
	}
};

type Options = {
  paths?: string[],
  moduleDirectory?: ModuleDirectory,
  extensions?: string[],
  alias?: Alias,
  transform?: Transform,
};

type ResolveOptions = {
  alias?: Alias,
  moduleDirectory?: string[],
  extensions?: string[],
  plugins?: any[]
}
export function createResolver(trackedFiles: string[], extensions: string[]): typeof resolve {
  return (mod: string, source: string, options: ResolveOptions) => resolve(mod, source, {
    ...options,
    plugins: [new FileListIncludesPlugin(trackedFiles, extensions)]
  });
};

async function dependenciesInRevision(revision: string, exports: Exports, options: Options) {
  const { extensions = DEFAULT_EXTENSIONS, paths, transform, moduleDirectory, alias } = options;
  const trackedFiles = getTrackedFiles(revision, paths);
  const trackedFilesAbsolute = trackedFiles.map(getAbsolutePath);
  return await filterDependents(trackedFilesAbsolute, exports, {
    moduleDirectory,
    alias,
    extensions,
    loader: createLoader(revision, transform),
    resolver: createResolver(trackedFiles, extensions)
  });
}

async function gitChangesAffected(
  commit: string, 
  options: Options = {} 
) {
  const { extensions = DEFAULT_EXTENSIONS, transform } = options;
  const [exportMap, staleExportMap] = await getDiffExportMap(commit, { 
    transform,
    extensions
  });
  const currentAffected = await dependenciesInRevision(commit, exportMap, options);
  const staleAffected = await dependenciesInRevision(`${commit}~1`, staleExportMap, options);
  /** @todo dedupe */
  return currentAffected.concat(staleAffected);
}
export default gitChangesAffected;