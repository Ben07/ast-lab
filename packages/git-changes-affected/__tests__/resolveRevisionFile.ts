import path from 'path';
import { createResolver, getTrackedFiles } from '../src';

describe('createResolver()', () => {
  test('can work', async () => {
    expect(await createResolver(
      getTrackedFiles('953d1596ff7df0016edb56d54195b445929c7c6e'),
      ['.js', '.ts']
    )(
      '../test/filterDependents.test',
      path.resolve(__dirname, '../../get-dependencies/src/index')
    )).toMatchSnapshot();
  });
});