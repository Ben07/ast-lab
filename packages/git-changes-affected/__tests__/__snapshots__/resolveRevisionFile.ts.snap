// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResolver() can work 1`] = `"/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/filterDependents.test.ts"`;
